app/services/idempotency/store.rb

```diff
@@ -0,0 +1,64 @@
+# frozen_string_literal: true
+
+require 'redis'
+
+module Idempotency
+  class Store
+    # NB the caching vs conflict logic is:
+    # *  re-using the idempotency key with the same request for 10 mins will return the same cached response
+    # *  re-using the idempotency key with the same request after 10 mins to 12 hours will raise a conflict error
+    # *  re-using the idempotency key with a different request will raise a conflict error, for 12 hours
+
+    CACHE_RESPONSE_TTL = 600 # 10 minutes
+    CONFLICT_TTL = 43_200 # 12 hours
+
+    attr_reader :idempotency_key, :conflict_key, :cache_response_key
+
+    def initialize(request)
+      # NB: we need to do a case-insensitive match for IDEMPOTENCY_KEY; rails does not do this automatically if the key contains an underscore
+      @idempotency_key = request.headers.find { |key, _v| key =~ /\AIDEMPOTENCY[\_\-]KEY\Z/i }&.last
+      if @idempotency_key.present?
+        @conflict_key = "conf|#{idempotency_key}"
+        @cache_response_key = "resp|#{idempotency_key}|#{request_hash(request)}"
+      end
+    end
+
+    def get
+      return if @idempotency_key.blank?
+
+      # Return the cached response if it matches the idempotency key and request
+      cached_response = redis.hgetall(cache_response_key)
```

Comment:
Does this really necessitate dropping down to native Redis? Rails has very good caching wrappers for this stuff and built in support for testing cache plus enabling/disabling caching in local development. If we can use those API's rather than direct Redis it would be nicer.

Reply:
I did look at the rails caching to start off with, but for one reason or another (I forget 🤔 ) opted for native redis. I can revisit this though and get back to you.

Reply:
So, I have thought further on this, and when I was thinking about the rails caching layer, I naively only considered a file-based approach: which wouldn't work as we have multiple pods. However a rails cache backed by redis (probably) will work, so I will see if I can implement that here.

Reply:
I suspect with Rails caching in the controller concern you likely won't need this class at all. And yes, we already have Redis setup and integrated in production for Sidekiq - maybe we need a few config lines in `production.rb` to enable it for caching too.